src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/CreditManager.cs

```diff
@@ -124,10 +120,19 @@ public void Dispose()
                 {
                     while (_waiters.TryDequeue(out Waiter waiter))
                     {
-                        waiter.TaskCompletionSource.TrySetException(new ObjectDisposedException(nameof(CreditManager)));
+                        waiter.TrySetException(CreateObjectDisposedException(forActiveWaiter: true));
                     }
                 }
             }
         }
+
+        private ObjectDisposedException CreateObjectDisposedException(bool forActiveWaiter) => forActiveWaiter ?
+            new ObjectDisposedException($"{nameof(CreditManager)}:{_owner.GetType().Name}:{_name}", SR.net_http_disposed_while_in_use) :
+            new ObjectDisposedException($"{nameof(CreditManager)}:{_owner.GetType().Name}:{_name}");
+
+        private sealed class Waiter : TaskCompletionSourceWithCancellation<int>
```

Comment:
Why is this better than the previous approach where Waiter was a struct? It seems fine, but I'd like to understand better...

Reply:
It's a bit smaller (and IMHO simpler).  The array in the queue would need to store 16 bytes for each entry, with 4 bytes wasted per entry in padding, whereas this approach doesn't have that.  This also maps better to what TaskCompletionSourceWithCancellation already does, which is storing the additional data right onto the single class instance.  And it makes the consuming code simpler, in that rather than doing waiter.TaskCompletionSource.SetCanceled it can just do waiter.SetCanceled.

Reply:
Ok, sounds good to me